### PR TITLE
[MathUtils] Add stable norm

### PR DESCRIPTION
**Pull request by oberbichler**
`StableNorm` is similar to the `Norm`, but reduces the risk of underflow and overflow.

### CODE REVIEW

File: kratos/utilities/math_utils.h [hunk 1]
```diff
@@ -827,6 +827,47 @@ class MathUtils
...
+
+        for (auto it = a.begin(); it != a.end(); ++it) {
+            TDataType x = *it;
+
+            if (x != 0) {
+                const TDataType abs_x = std::abs(x);
+
+                if (max < abs_x) {
+                    x = max / abs_x;
+                    sqr_sum_scaled = sqr_sum_scaled * (x * x) + 1.0;
```
Comment (loumalouomega) [inline kratos/utilities/math_utils.h]: Where the +1..0 comes from?, it is not in the link for LAPACK you added (thanks :D )

**Review by loumalouomega — CHANGES_REQUESTED** [review]
Can you clarify the origin of the formulation? (+1 I mean)

**Review by loumalouomega — DISMISSED** [review]
Ok, now I see that they wrote "one" in the code. a little bit obscure...

File: kratos/utilities/math_utils.h (hunk 1 above)
Comment (loumalouomega) [reply kratos/utilities/math_utils.h]: They wrote one avoid define multiple times 1.0d...

Comment (oberbichler) [reply kratos/utilities/math_utils.h]: The entries of the value are scaled according to the maximum. This line is executed when `x[i]` is the new maximum. So `x[i] / max = x[i] / x[i] = 1`. But I maybe should not reuse `x` in this scenario...

**Comment by oberbichler** [comment]
Yes... Please tell me if you know the reason why.

I am also wondering if we have some practical examples where the simple norm overflows. Just found some solvers where there are using the stable norm for residual calculations.

**Comment by oberbichler** [comment]
just used some better variable names...

**Review by loumalouomega — APPROVED** [review]

**Pull request merged**